projects/extension/src/background/ConnectionManagerWithHealth.ts

```diff
@@ -301,6 +335,19 @@ export class ConnectionManagerWithHealth<SandboxId> {
                   }
                 }
               }
+              // If jsonRpcMessageId starts with "extension:" then this message is a response to one
+              // initiated from extension's pages (Options page). It needs to have a different
+              // return type to the extension, in order to identify it and retrieve information from
+              // created Map and save to the localStorage
+            } else if (jsonRpcMessageId.startsWith("extension:")) {
+              jsonRpcMessage.id = JSON.parse(
+                (jsonRpcMessage.id as string).slice("extension:".length),
+              )
+              return {
+                origin: "connection-manager",
+                type: "extension-config",
+                jsonRpcMessage: JSON.stringify(jsonRpcMessage),
+              }
```

Comment:
...and then here you use the `Map`, instead of returning a message.

Reply:
Im not sure what you mean. This is needed, in order to "inform" `index` that rpc was returned and returning `type: "extension-config"` is also needed in order to identify that this is not any other RPC and initiate the `verifyBootnode` listeners.. Id and error that are returning from the rpc are needed for mapping the outcome thus jsonRpcMessage must be returned to `index` (error message should appear to the UI).

Reply:
Right now you call `validateBootnode` you have the guarantee to later receive back a JSON-RPC response of type `extension-config` (why "extension-config"?) containing an id that you must then pass to `getBootnode`.
I don't know if that makes sense to you, but I consider this a spaghetti mess.

What I'm saying is that you should get rid of this idea of returning an `extension-config` message and of `getBootnode`, and instead pass some kind of callback to `validateBootnode` that is called at this point in the code (where we're commenting right now).

Reply:
Right now you call validateBootnode, and you then have the guarantee to later receive a JSON-RPC response of type extension-config (why "extension-config"?) containing an id that you must then pass to getBootnode.
I don't know if that makes sense to you, probably yes since you've design this, but I consider this a spaghetti mess.

What I'm saying is that you should get rid of this idea of returning an extension-config message and of getBootnode, and instead put some kind of callback inside the `Map`, that is called at this point in the code (where we're commenting right now).

Then you can change `validateBootnode`. The callback that is put in the map could be provided to `validateBootnode`, but a `Promise` would IMO be even better.